lms/product/plugin/course_copy.py

```diff
@@ -0,0 +1,74 @@
+from lms.models import File
+from lms.services.exceptions import CanvasFileNotFoundInCourse
+
+
+class CourseCopyPlugin:
+    file_type = None
+
+    def assert_file_in_course(self, course_id, file_id):
+        """Raise if the current user can't see file_id in course_id."""
+        raise NotImplementedError()
+
+    def find_matching_file_in_course(self, original_file_id, new_course_id):
+        try:
+            # Get the current (copied) courses files, that will have the side effect of storing files in the DB
+            _ = self._store_new_course_files(new_course_id)
+        except ExternalRequestError:
+            # We might not have access to use the API for that endpoint.
+            # That will depend on our role and the course's permissions settings.
+            # We will continue anyway, maybe the files of the new course are already in the DB
+            # after an instructor launched corrected the issue.
+            pass
+
+        # We get the original file record from the DB
+        file = self._file_service.get(original_file_id, type_=self.file_type)
+        if not file:
+            # That file must have been recorded when the original assignment was configured.
+            # If we can't find that one something odd is going on, stop here.
+            return
+
+        # Now we'll try to find a matching file in the DB in the new course
+        # We might have a record of this because we just called `_store_new_course_files` as the current user
+        # or another user might have done it before for us.
+        new_file = (
```

Comment:
This query won't be just pasted here in the final version, it will over over `file_service`